packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1459 @@
+        uint8 noSubmissionCount;
+        mapping(address => KroDelegator) kroDelegators;
+        mapping(address => KghDelegator) kghDelegators;
+    }
+
+    /**
+     * @notice The numerator for the boosted reward.
+     */
+    uint128 public constant BOOSTED_REWARD_NUMERATOR = 40;
+
+    /**
+     * @notice The denominator for the boosted reward.
+     */
+    uint128 public constant BOOSTED_REWARD_DENOM = 100;
+
+    /**
+     * @notice The denominator for per-mille used at slashing and initialization of undelegation.
+     */
+    uint128 public constant PER_MILLE_DENOM = 1000;
+
+    /**
+     * @notice Virtual KRO amount per KGH.
+     */
+    uint128 public constant VKRO_PER_KGH = 100e18;
+
+    /**
+     * @notice The numerator of the tax.
+     */
+    uint128 public constant TAX_NUMERATOR = 20;
+
+    /**
+     * @notice The denominator of the tax.
+     */
+    uint128 public constant TAX_DENOMINATOR = 100;
+
+    /**
+     * @notice Decimals offset for the KRO and KGH shares.
+     */
+    uint128 public constant DECIMAL_OFFSET = 10 ** 6;
+
+    /**
+     * @notice The maximum number of KROs that can be undelegated at once.
+     */
+    uint256 public constant MAX_DELEGATE = 50_000_000 * 10 ** 18;
+
+    /**
+     * @notice Address of the KRO token contract.
+     */
+    IERC20 public immutable TOKEN;
+
+    /**
+     * @notice Address of the L2OutputOracle contract. Can be updated via upgrade.
+     */
+    L2OutputOracle public immutable L2_ORACLE;
+
+    /**
+     * @notice Address of the KGH token contract.
+     */
+    IERC721 public immutable KGH;
+
+    /**
+     * @notice Address of the KGHManager contract.
+     */
+    IKGHManager public immutable KGH_MANAGER;
+
+    /**
+     * @notice The address of the SecurityCouncil contract. Can be updated via upgrade.
+     */
+    address public immutable SECURITY_COUNCIL;
+
+    /**
+     * @notice The max number of reward recipients when distributing rewards.
+     */
+    uint128 public immutable MAX_REWARD_RECIPIENTS;
+
+    /**
+     * @notice Amount of base reward for the validator.
+     */
+    uint128 public immutable BASE_REWARD;
+
+    /**
+     * @notice The numerator of the slashing rate. It will be divided by PER_MILLE_DENOM.
+     */
+    uint128 public immutable SLASHING_RATE_NUMERATOR;
+
+    /**
+     * @notice Minimum amount to slash.
+     */
+    uint128 public immutable MIN_SLASHING_AMOUNT;
+
+    /**
+     * @notice Minimum amount to register as a validator.
+     */
+    uint128 public immutable MIN_REGISTER_AMOUNT;
+
+    /**
+     * @notice Minimum amount to start submitting outputs.
+     */
+    uint128 public immutable MIN_START_AMOUNT;
+
+    /**
+     * @notice Withdrawal delay for the undelegation.
+     */
+    uint256 public immutable WITHDRAWAL_DELAY;
+
+    /**
+     * @notice Weighted tree to store and calculate the probability to be selected as an output submitter.
+     */
+    BalancedWeightTree.Tree internal _weightTree;
```

Comment:
The weightTree is a data structure, so I recommend to use the name for this `_validatorTree`.
```suggestion
    BalancedWeightTree.Tree internal _validatorTree;
```